### PR TITLE
Add support to zero db to support simple groups.

### PR DESCRIPTION
**Pull request by litghost**
Previously these kinds of zero groups would require encoding the
final bits, rather than tags.  This is extends the dbfixup to
construct groups via groups of tags, rather than groups of bits.

This is PR adds support to enable https://github.com/SymbiFlow/prjxray/issues/817 to be completed.

### CODE REVIEW

File: utils/dbfixup.py [hunk 1]
```diff
@@ -51,57 +51,153 @@ def bits_str(bits):
...
-            zdb, allzero_tag = zdb.split(",")
-
-        if "|" in zdb:
-            a, b = zdb.split("|")
-            a = a.split()
-            b = b.split()
-        else:
-            a = zdb.split()
-            b = a
+        Zero db format examples:
```
Comment (mithro) [inline utils/dbfixup.py]: So, I know you didn't come up with this, but I do find this syntax quite confusing....

File: utils/dbfixup.py [hunk 2]
```diff
@@ -51,57 +51,153 @@ def bits_str(bits):
...
 
-        bitmatch = False
-        for bit in a:
-            if bit in bits:
-                bitmatch = True
+        Ex: 01_02 04_05
```
Comment (mithro) [inline utils/dbfixup.py]: Doesn't this mean `01_02` and `04_05` part of a group even if neither of these bits are discovered to be connected?

File: utils/dbfixup.py [hunk 3]
```diff
@@ -51,57 +51,153 @@ def bits_str(bits):
...
-            continue
+        Ex: 01_02 04_05|07_08 10_11
+        If any bits from the first group occur,
+        default bits in the second group to zero
+
+        Ex: 01_02 04_05,ALL_ZERO
+        ALL_ZERO is an enum that is part of the group but is all 0
+        It must have 0 candidates
+
+        Ex: CLB.SLICE_X0.CLKINV ^ CLB.SLICE_X0.NOCLKINV
```
Comment (mithro) [inline utils/dbfixup.py]: I think this would be a more correct syntax?
```
Ex: ^CLB.SLICE_X0.CLKINV,NOCLKINV
```

Comment (litghost) [reply utils/dbfixup.py]: No?

File: utils/dbfixup.py (hunk 1 above)
Comment (litghost) [reply utils/dbfixup.py]: Would you like to be more specific about what is confusing, it doesn't seem to bad.  This comment is not actionable, and I'm not planning on changing every dbf file over a vague comment.

File: utils/dbfixup.py (hunk 2 above)
Comment (litghost) [reply utils/dbfixup.py]: Your question doesn't make sense, rephrase?

Comment (mithro) [reply utils/dbfixup.py]: If I understand things correctly, then when starting with the following output;
```
01_02 BLAH.A
04_05 BLAH.B
```

----

Then with `.dbf` file with the following contents;
```
01_02 04_05
```
The output will make sure that `BLAH.A` and `BLAH.B` are used together as a group and can never be set at the same time. IE The actual bit pattern should be,
```
 01_02 !04_05 BLAH.A
!01_02  04_05 BLAH.B
```

----

Then with a `.dbf` file with the following contents;
```
01_02 04_05,ALL_ZERO
```

You get the following bit patterns,
```
 01_02 !04_05 BLAH.A
!01_02  04_05 BLAH.B
!01_02 !04_05 BLAH.ALL_ZERO
```

----

Then with a `.dbf` file with the following contents;
```
01_02 04_05|07_08 10_11
```

You get the following bit patterns,
```
 01_02 !04_05 !07_08 !10_11 BLAH.A
!01_02  04_05 !07_08 !10_11 BLAH.B
!01_02 !04_05  07_08 !10_11 ????
!01_02 !04_05 !07_08  10_11 ????
```

----

Is my understanding incorrect?

Comment (litghost) [reply utils/dbfixup.py]: > Then with a `.dbf` file with the following contents;
> 
> ```
> 01_02 04_05|07_08 10_11
> ```
>
>  01_02 !04_05 !07_08 !10_11 BLAH.A
> !01_02  04_05 !07_08 !10_11 BLAH.B
> !01_02 !04_05  07_08 !10_11 ????
> !01_02 !04_05 !07_08  10_11 ????

No, the result is:
```
01_02 BLAH.A
04_05 BLAH.B
```

**Comment by litghost** [comment]
Red CI is due to 044 failing, re-running for now.

File: utils/dbfixup.py (hunk 2 above)
Comment (mithro) [reply utils/dbfixup.py]: What about the others?

Comment (litghost) [reply utils/dbfixup.py]: You mean ``07_08`` and ``10_11``?  Nothing special?

**Comment by HackerFoo** [comment]
Is there an explanation of the syntax change? Is it:
```
FOO.A FOO.B 01_02 03_04
```

to mean "if `FOO.A` and `FOO.B`, then `01_02` and `03_04`"?

If so, maybe we could add an arrow to match a common notation:

```
FOO.A FOO.B -> 01_02 03_04

# could also be written
FOO.A FOO.B -> 01_02
FOO.A FOO.B -> 03_04
```

If it's `01_02` or `03_04`, we might need a SAT solver.

**Pull request merged**